File: app/src/main/java/com/riprog/launcher/WidgetManager.java
package com.riprog.launcher;

import android.app.AlertDialog;
import android.app.Dialog;
import android.appwidget.AppWidgetHost;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProviderInfo;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.os.Build;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class WidgetManager {
    private final MainActivity activity;
    private final SettingsManager settingsManager;
    private final AppWidgetManager appWidgetManager;
    private final AppWidgetHost appWidgetHost;
    private final GridManager gridManager;
    private final ExecutorService widgetPreviewExecutor = Executors.newFixedThreadPool(4);

    public WidgetManager(MainActivity activity, SettingsManager settingsManager, AppWidgetManager appWidgetManager, AppWidgetHost appWidgetHost) {
        this.activity = activity;
        this.settingsManager = settingsManager;
        this.appWidgetManager = appWidgetManager;
        this.appWidgetHost = appWidgetHost;
        this.gridManager = new GridManager(settingsManager.getColumns());
    }

    public void pickWidget(float lastGridCol, float lastGridRow) {
        if (appWidgetManager == null) return;
        List<AppWidgetProviderInfo> providers = appWidgetManager.getInstalledProviders();
        if (providers == null) return;

        Map<String, List<AppWidgetProviderInfo>> grouped = new HashMap<>();
        for (AppWidgetProviderInfo info : providers) {
            String pkg = info.provider.getPackageName();
            if (!grouped.containsKey(pkg)) grouped.put(pkg, new ArrayList<>());
            grouped.get(pkg).add(info);
        }

        List<String> packages = new ArrayList<>(grouped.keySet());
        Collections.sort(packages, (a, b) -> activity.getAppName(a).compareToIgnoreCase(activity.getAppName(b)));

        Dialog dialog = new Dialog(activity, android.R.style.Theme_DeviceDefault_NoActionBar_Fullscreen);
        if (dialog.getWindow() != null) {
            dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
            ThemeUtils.applyWindowBlur(dialog.getWindow(), settingsManager.isLiquidGlass());
        }

        FrameLayout root = new FrameLayout(activity);
        root.setBackground(ThemeUtils.getGlassDrawable(activity, settingsManager, 0f));

        LinearLayout container = new LinearLayout(activity);
        container.setOrientation(LinearLayout.VERTICAL);
        root.addView(container);

        int adaptiveColor = ThemeUtils.getAdaptiveColor(activity, settingsManager, true);
        int secondaryColor = (adaptiveColor & 0x00FFFFFF) | 0x80000000;

        LinearLayout titleLayout = new LinearLayout(activity);
        titleLayout.setOrientation(LinearLayout.HORIZONTAL);
        titleLayout.setGravity(Gravity.CENTER_VERTICAL);
        titleLayout.setPadding(0, 0, 0, dpToPx(32));

        ImageView titleIcon = new ImageView(activity);
        titleIcon.setImageResource(R.drawable.ic_widgets);
        titleIcon.setColorFilter(adaptiveColor);
        titleLayout.addView(titleIcon, new LinearLayout.LayoutParams(dpToPx(32), dpToPx(32)));

        TextView title = new TextView(activity);
        title.setText(R.string.title_pick_widget);
        title.setTextSize(32f);
        title.setTypeface(Typeface.create("sans-serif-light", Typeface.NORMAL));
        title.setTextColor(adaptiveColor);
        LinearLayout.LayoutParams titleParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        titleParams.leftMargin = dpToPx(16);
        titleLayout.addView(title, titleParams);
        container.addView(titleLayout);

        ScrollView scrollView = new ScrollView(activity);
        scrollView.setVerticalScrollBarEnabled(false);
        LinearLayout itemsContainer = new LinearLayout(activity);
        itemsContainer.setOrientation(LinearLayout.VERTICAL);
        scrollView.addView(itemsContainer);
        container.addView(scrollView);

        for (String pkg : packages) {
            TextView header = new TextView(activity);
            header.setText(activity.getAppName(pkg));
            header.setTextSize(12f);
            header.setTypeface(null, Typeface.BOLD);
            header.setTextColor(secondaryColor);
            header.setAllCaps(true);
            header.setPadding(0, dpToPx(24), 0, dpToPx(12));
            itemsContainer.addView(header);

            List<AppWidgetProviderInfo> infos = grouped.get(pkg);
            if (infos != null) {
                for (AppWidgetProviderInfo info : infos) {
                    LinearLayout card = new LinearLayout(activity);
                    card.setOrientation(LinearLayout.HORIZONTAL);
                    card.setGravity(Gravity.CENTER_VERTICAL);
                    card.setPadding(dpToPx(16), dpToPx(16), dpToPx(16), dpToPx(16));
                    card.setClickable(true);
                    card.setFocusable(true);

                    GradientDrawable cardBg = new GradientDrawable();
                    boolean isNight = (activity.getResources().getConfiguration().uiMode & android.content.res.Configuration.UI_MODE_NIGHT_MASK) ==
                            android.content.res.Configuration.UI_MODE_NIGHT_YES;
                    int cardColor = settingsManager.isLiquidGlass() ? 0x1AFFFFFF : (isNight ? 0x1AFFFFFF : 0x0D000000);
                    cardBg.setColor(cardColor);
                    cardBg.setCornerRadius(dpToPx(16));
                    card.setBackground(cardBg);

                    LinearLayout.LayoutParams cardLp = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
                    cardLp.bottomMargin = dpToPx(12);
                    itemsContainer.addView(card, cardLp);

                    ImageView preview = new ImageView(activity);
                    preview.setScaleType(ImageView.ScaleType.FIT_CENTER);
                    float density = activity.getResources().getDisplayMetrics().density;

                    int cellWidth = gridManager.getCellWidth(activity.homeView.getWidth() > 0 ? activity.homeView.getWidth() : activity.getResources().getDisplayMetrics().widthPixels);
                    int cellHeight = gridManager.getCellHeight(activity.homeView.getHeight() > 0 ? activity.homeView.getHeight() : activity.getResources().getDisplayMetrics().heightPixels);

                    float sX = gridManager.calculateSpanX(info.minWidth * density, cellWidth);
                    float sY = gridManager.calculateSpanY(info.minHeight * density, cellHeight);

                    if (!settingsManager.isFreeformHome()) {
                        sX = Math.max(1, Math.round(sX));
                        sY = Math.max(1, Math.round(sY));
                    }
                    final float spanX = sX;
                    final float spanY = sY;

                    widgetPreviewExecutor.execute(() -> {
                        try {
                            Drawable previewDrawable = info.loadPreviewImage(activity, 0);
                            if (previewDrawable == null) previewDrawable = info.loadIcon(activity, 0);
                            final Drawable finalDrawable = previewDrawable;
                            activity.runOnUiThread(() -> preview.setImageDrawable(finalDrawable));
                        } catch (Exception ignored) { }
                    });
                    card.addView(preview, new LinearLayout.LayoutParams(dpToPx(64), dpToPx(64)));

                    LinearLayout textLayout = new LinearLayout(activity);
                    textLayout.setOrientation(LinearLayout.VERTICAL);
                    LinearLayout.LayoutParams textLp = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
                    textLp.leftMargin = dpToPx(16);
                    card.addView(textLayout, textLp);

                    TextView label = new TextView(activity);
                    label.setText(info.loadLabel(activity.getPackageManager()));
                    label.setTextColor(adaptiveColor);
                    label.setTextSize(16f);
                    label.setTypeface(null, Typeface.BOLD);
                    textLayout.addView(label);

                    TextView size = new TextView(activity);
                    size.setText(activity.getString(R.string.widget_size_format, (int)Math.ceil(spanX), (int)Math.ceil(spanY)));
                    size.setTextSize(12f);
                    size.setTextColor(secondaryColor);
                    textLayout.addView(size);

                    card.setOnClickListener(v -> Toast.makeText(activity, "Long press to drag widget", Toast.LENGTH_SHORT).show());
                    card.setOnLongClickListener(v -> {
                        dialog.dismiss();
                        activity.startNewWidgetDrag(info, (int)spanX, (int)spanY);
                        return true;
                    });
                }
            }
        }

        ImageView closeBtn = new ImageView(activity);
        closeBtn.setImageResource(android.R.drawable.ic_menu_close_clear_cancel);
        closeBtn.setColorFilter(adaptiveColor);
        closeBtn.setAlpha(0.6f);
        closeBtn.setOnClickListener(v -> dialog.dismiss());
        FrameLayout.LayoutParams closeLp = new FrameLayout.LayoutParams(dpToPx(48), dpToPx(48), Gravity.TOP | Gravity.END);
        closeLp.topMargin = dpToPx(16);
        closeLp.rightMargin = dpToPx(16);
        root.addView(closeBtn, closeLp);

        ViewCompat.setOnApplyWindowInsetsListener(root, (v, insets) -> {
            androidx.core.graphics.Insets bars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            container.setPadding(dpToPx(24), bars.top + dpToPx(64), dpToPx(24), bars.bottom + dpToPx(24));
            closeLp.topMargin = bars.top + dpToPx(16);
            closeBtn.setLayoutParams(closeLp);
            return insets;
        });

        dialog.setContentView(root);
        dialog.show();
    }

    private int dpToPx(float dp) {
        return (int) android.util.TypedValue.applyDimension(
            android.util.TypedValue.COMPLEX_UNIT_DIP, dp, activity.getResources().getDisplayMetrics());
    }
}

File: app/src/main/java/com/riprog/launcher/ui/home/manager/WidgetManager.kt
package com.riprog.launcher.ui.home.manager

import android.app.AlertDialog
import android.app.Dialog
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProviderInfo
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.GradientDrawable
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.riprog.launcher.ui.home.MainActivity
import com.riprog.launcher.R
import com.riprog.launcher.data.model.HomeItem
import com.riprog.launcher.ui.home.HomeView
import com.riprog.launcher.data.local.prefs.LauncherPreferences
import com.riprog.launcher.ui.common.ThemeUtils
import java.util.concurrent.Executors

class WidgetManager(
    private val activity: MainActivity,
    private val settingsManager: LauncherPreferences,
    private val appWidgetManager: AppWidgetManager?,
    private val appWidgetHost: android.appwidget.AppWidgetHost?
) {
    private val gridManager = GridManager(settingsManager.columns)
    private val widgetPreviewExecutor = Executors.newFixedThreadPool(4)

    fun pickWidget(lastGridCol: Float, lastGridRow: Float) {
        val am = appWidgetManager ?: return
        val providers = am.installedProviders ?: return
        val grouped = mutableMapOf<String, MutableList<AppWidgetProviderInfo>>()
        for (info in providers) {
            val pkg = info.provider.packageName
            grouped.getOrPut(pkg) { mutableListOf() }.add(info)
        }

        val packages = grouped.keys.sortedWith { a, b -> activity.getAppName(a).compareTo(activity.getAppName(b), true) }

        val dialog = Dialog(activity, android.R.style.Theme_DeviceDefault_NoActionBar_Fullscreen)
        dialog.window?.let {
            it.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            ThemeUtils.applyWindowBlur(it, settingsManager.isLiquidGlass)
        }

        val root = FrameLayout(activity).apply { background = ThemeUtils.getGlassDrawable(activity, settingsManager, 0f) }
        val container = LinearLayout(activity).apply { orientation = LinearLayout.VERTICAL }
        root.addView(container)

        val adaptiveColor = ThemeUtils.getAdaptiveColor(activity, settingsManager, true)
        val secondaryColor = (adaptiveColor and 0x00FFFFFF) or 0x80000000.toInt()

        val titleLayout = LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            setPadding(0, 0, 0, dpToPx(32))
        }

        val titleIcon = ImageView(activity).apply {
            setImageResource(R.drawable.ic_widgets)
            setColorFilter(adaptiveColor)
        }
        titleLayout.addView(titleIcon, LinearLayout.LayoutParams(dpToPx(32), dpToPx(32)).apply { rightMargin = dpToPx(16) })

        val title = TextView(activity).apply {
            setText(R.string.title_pick_widget)
            textSize = 32f
            typeface = Typeface.create("sans-serif-light", Typeface.NORMAL)
            setTextColor(adaptiveColor)
        }
        titleLayout.addView(title)
        container.addView(titleLayout)

        val scrollView = ScrollView(activity).apply { isVerticalScrollBarEnabled = false }
        val itemsContainer = LinearLayout(activity).apply { orientation = LinearLayout.VERTICAL }
        scrollView.addView(itemsContainer)
        container.addView(scrollView)

        for (pkg in packages) {
            val header = TextView(activity).apply {
                text = activity.getAppName(pkg)
                textSize = 12f
                setTypeface(null, Typeface.BOLD)
                setTextColor(secondaryColor)
                isAllCaps = true
                setPadding(0, dpToPx(24), 0, dpToPx(12))
            }
            itemsContainer.addView(header)

            grouped[pkg]?.let { infos ->
                for (info in infos) {
                    val card = LinearLayout(activity).apply {
                        orientation = LinearLayout.HORIZONTAL
                        gravity = Gravity.CENTER_VERTICAL
                        setPadding(dpToPx(16), dpToPx(16), dpToPx(16), dpToPx(16))
                        isClickable = true
                        isFocusable = true
                    }

                    val cardBg = GradientDrawable()
                    val isNight = (activity.resources.configuration.uiMode and android.content.res.Configuration.UI_MODE_NIGHT_MASK) == android.content.res.Configuration.UI_MODE_NIGHT_YES
                    val cardColor = if (settingsManager.isLiquidGlass) 0x1AFFFFFF else (if (isNight) 0x1AFFFFFF else 0x0D000000)
                    cardBg.setColor(cardColor)
                    cardBg.cornerRadius = dpToPx(16).toFloat()
                    card.background = cardBg

                    itemsContainer.addView(card, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT).apply { bottomMargin = dpToPx(12) })

                    val preview = ImageView(activity).apply { scaleType = ImageView.ScaleType.FIT_CENTER }
                    val density = activity.resources.displayMetrics.density
                    val hv = activity.getHomeView()
                    val grid = hv?.gridManager ?: gridManager
                    val availW = if (hv != null && hv.width > 0) hv.width - hv.paddingLeft - hv.paddingRight else activity.resources.displayMetrics.widthPixels
                    val availH = if (hv != null && hv.height > 0) hv.height - hv.paddingTop - hv.paddingBottom else activity.resources.displayMetrics.heightPixels
                    val cellWidth = grid.getCellWidth(availW)
                    val cellHeight = grid.getCellHeight(availH)
                    var sX = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S && info.targetCellWidth > 0) {
                        info.targetCellWidth.toFloat()
                    } else {
                        grid.calculateSpanX(info.minWidth * density, cellWidth)
                    }
                    var sY = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S && info.targetCellHeight > 0) {
                        info.targetCellHeight.toFloat()
                    } else {
                        grid.calculateSpanY(info.minHeight * density, cellHeight)
                    }

                    if (!settingsManager.isFreeformHome) {
                        sX = Math.round(sX).coerceAtLeast(1).toFloat()
                        sY = Math.round(sY).coerceAtLeast(1).toFloat()
                    }
                    val spanX = sX
                    val spanY = sY

                    widgetPreviewExecutor.execute {
                        try {
                            val previewDrawable = info.loadPreviewImage(activity, 0) ?: info.loadIcon(activity, 0)
                            activity.runOnUiThread { preview.setImageDrawable(previewDrawable) }
                        } catch (ignored: Exception) { }
                    }
                    card.addView(preview, LinearLayout.LayoutParams(dpToPx(64), dpToPx(64)).apply { rightMargin = dpToPx(16) })

                    val textLayout = LinearLayout(activity).apply { orientation = LinearLayout.VERTICAL }
                    textLayout.addView(TextView(activity).apply {
                        text = info.label
                        setTextColor(adaptiveColor)
                        textSize = 16f
                        setTypeface(null, Typeface.BOLD)
                    })
                    textLayout.addView(TextView(activity).apply {
                        text = activity.getString(R.string.widget_size_format, Math.ceil(spanX.toDouble()).toInt(), Math.ceil(spanY.toDouble()).toInt())
                        textSize = 12f
                        setTextColor(secondaryColor)
                    })
                    card.addView(textLayout)

                    card.setOnClickListener {
                        dialog.dismiss()
                        val appWidgetId = appWidgetHost!!.allocateAppWidgetId()
                        val currentPage = activity.getHomeView()?.getCurrentPage() ?: 0
                        activity.setPendingWidgetParams(lastGridCol, lastGridRow, spanX, spanY, currentPage)
                        if (am.bindAppWidgetIdIfAllowed(appWidgetId, info.provider)) {
                            activity.createWidgetAt(appWidgetId, lastGridCol, lastGridRow, spanX, spanY, currentPage)
                        } else {
                            val intent = Intent(AppWidgetManager.ACTION_APPWIDGET_BIND).apply {
                                putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
                                putExtra(AppWidgetManager.EXTRA_APPWIDGET_PROVIDER, info.provider)
                            }
                            activity.startActivityForResult(intent, MainActivity.REQUEST_PICK_APPWIDGET)
                        }
                    }
                }
            }
        }

        val closeBtn = ImageView(activity).apply {
            setImageResource(android.R.drawable.ic_menu_close_clear_cancel)
            setColorFilter(adaptiveColor)
            alpha = 0.6f
            setOnClickListener { dialog.dismiss() }
        }
        root.addView(closeBtn, FrameLayout.LayoutParams(dpToPx(48), dpToPx(48), Gravity.TOP or Gravity.RIGHT).apply { topMargin = dpToPx(16); rightMargin = dpToPx(16) })

        root.setOnApplyWindowInsetsListener { _, insets ->
            val top = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) insets.getInsets(android.view.WindowInsets.Type.systemBars()).top else insets.systemWindowInsetTop
            val bottom = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) insets.getInsets(android.view.WindowInsets.Type.systemBars()).bottom else insets.systemWindowInsetBottom
            container.setPadding(dpToPx(24), top + dpToPx(64), dpToPx(24), bottom + dpToPx(24))
            val clp = closeBtn.layoutParams as FrameLayout.LayoutParams
            clp.topMargin = top + dpToPx(16)
            closeBtn.layoutParams = clp
            insets
        }

        dialog.setContentView(root)
        dialog.show()
    }

    fun showWidgetOptions(item: HomeItem, hostView: View) {
        val options = arrayOf(activity.getString(R.string.action_resize), activity.getString(R.string.action_remove))
        val dialog = AlertDialog.Builder(activity, android.R.style.Theme_DeviceDefault_Dialog_Alert)
            .setItems(options) { _, which ->
                if (which == 0) showResizeDialog(item, hostView)
                else activity.removeHomeItem(item, hostView)
            }.create()
        dialog.show()
        dialog.window?.let {
            it.setBackgroundDrawable(ThemeUtils.getGlassDrawable(activity, settingsManager))
            ThemeUtils.applyWindowBlur(it, settingsManager.isLiquidGlass)
        }
    }

    fun showResizeDialog(item: HomeItem, hostView: View) {
        val sizes = arrayOf("1x1", "2x1", "2x2", "4x2", "4x1")
        val dialog = AlertDialog.Builder(activity, android.R.style.Theme_DeviceDefault_Dialog_Alert)
            .setTitle(R.string.title_resize_widget)
            .setItems(sizes) { _, which ->
                when (which) {
                    0 -> { item.spanX = 1f; item.spanY = 1f }
                    1 -> { item.spanX = 2f; item.spanY = 1f }
                    2 -> { item.spanX = 2f; item.spanY = 2f }
                    3 -> { item.spanX = 4f; item.spanY = 2f }
                    4 -> { item.spanX = 4f; item.spanY = 1f }
                }
                activity.getHomeView()?.updateViewPosition(item, hostView)
                activity.saveHomeState()
            }.create()
        dialog.show()
        dialog.window?.let {
            it.setBackgroundDrawable(ThemeUtils.getGlassDrawable(activity, settingsManager))
            ThemeUtils.applyWindowBlur(it, settingsManager.isLiquidGlass)
        }
    }

    private fun dpToPx(dp: Int): Int {
        return android.util.TypedValue.applyDimension(
            android.util.TypedValue.COMPLEX_UNIT_DIP, dp.toFloat(), activity.resources.displayMetrics
        ).toInt()
    }
}
